### ISSUE TITLE
Transmission fails to recover from byte error

### ISSUE THREAD

**Issue by alexwhittemore** [alexwoo1900/ymodem#15]
I've gotten transmission up and running with the included FileReceiver.py/FileSender.py, however the connection fails to gracefully handle byte errors (or my methods are wrong). 

For instance, modfiying `FileReceiver.py`:

```
    txcount = 0
    def read(size, timeout = 10):
        global txcount
        serial_io.timeout = timeout
        txcount +=1
        if txcount == 300:
            data = bytearray(serial_io.read(size))
            data[0] = b'0'
            return bytes(data)
        else:
            return serial_io.read(size)
```
Basically, if I just mangle the first byte of the 300th receive, I get

```
$python FileReceiver.py 
0 - sample.stl 26.00% [#############->.....................................]4.24s[Modem]: Read timeout!
[Receiver]: Received data timed out.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: The number of retransmissions has reached the maximum limit, abort and exit!
```
on the receive side and on the transmit side:
```
$python FileSender.py
0 - sample.stl 26.00% [#############->.....................................]4.23s[Sender]: No ACK from Receiver, 
preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: No ACK from Receiver, preparing to retransmit.
[Sender]: The number of retransmissions has reached the maximum limit, abort and exit!
```

I get similar behavior if I do the same mangling on the TX side.

**Comment by alexwhittemore** [comment]
Little update: Rather predictably, if I foul the connection by i.e. disconnecting one of the serial wires for a moment, the same behavior results - failure to re-sync and re-try the send.

**Comment by alexwhittemore** [comment]
And another: My method WAS wrong. The line `data[0] = b'0'` is invalid and throws an exception, which leads to the original "Read timeout!" (which just assumes any exception thrown in the `read()` function is a timeout). I corrected the fault injection to:

```python
def read(size, timeout = 10):
        global txcount
        serial_io.timeout = timeout
        txcount +=1
        if txcount == 300:
            data = serial_io.read(size)
            data = bytearray(data)
            data[0] = ord('0')
            return bytes(data)
        else:
            return serial_io.read(size)
```

That said, the failure persists, and the output is only slightly different. On the transmit side, same behavior as above. On the receive side, it's 
```
FileReceiver.py 
0 - sample.stl 26.00% [#############->.....................................]4.33s[Receiver]: Checksum failed.
[Receiver]: Send a request for retransmission.
[Receiver]: Wrong sequence, drop the whole packet.
[Receiver]: Send a request for retransmission.
```

**Comment by alexwhittemore** [comment]
More stream-of-consciousness while I try to solve this on my own:

I'm looking at this hunk, along with the other two similar sections that could possibly emit `No ACK from Receiver, preparing to retransmit.` https://github.com/alexwoo1900/ymodem/blob/8cf009b62f89ef1ae36416d9b4d3260404bf789b/ymodem/Socket.py#L279-L297C20

My first thought is that the logic looks possibly faulty. For one, `_read_and_wait` effectively tests for ANY character in the expected return, such that NAK or ACK could potentially match each other if a character or two somehow gets dropped. Would it make more sense to test `in_waiting()` for a buffer full to the expected length then read (3 or 4) characters at once, and test the whole string?

Then the next line that gets hit is 305, `c = self._read_and_wait([NAK, CRC, G, CAN], 60)`

Well, in the event that the receiver DID send a NAK, you'd only have AK left in the buffer at that test, right?

Anyway, still haven't found a smoking gun.

**Comment by alexwhittemore** [comment]
I relaxed my test a little. Now instead of mangling any old read, it'll ONLY mangle one byte of a long read so as to cause a CRC error, rather than potentially mangling a control signal. 
```python
    txcount = 0
    def read(size, timeout = 10):
        global txcount
        serial_io.timeout = timeout
        txcount +=1
        if txcount >100 and size>1:
            txcount = -100000
            data_in = serial_io.read(size)
            data = bytearray(data_in)
            data[0] = ord('0')
            log.debug(f"Munched data on {size} byte read, {data_in} to {bytes(data)}")
            return bytes(data)
        else:
            return serial_io.read(size)
```

This doesn't change the behavior substantially, just food for thought to make debugging easier. 

I also modified `Socket.py` lines 805-808:
```python
                    else:
                        # skip this packet
                        self.logger.warning(f"[Receiver]: Wrong sequence ({seq1}, {seq2}, {sequence}), drop the whole packet.")
                        self.read(packet_size + 1 + crc)
```
and added logging on 739:
``` python
                    self.logger.debug(f"Seq1: {seq1}")
                    self.logger.debug(f"Seq2: {seq2}")
```

This yields some interesting output:
```
2023-09-25 17:45:07,088 [Receiver]: <- Data packet 21
0 - sample.stl 8.00% [####->..............................................]1.20s2023-09-25 17:45:07,141 [Receiver]: ACK ->
2023-09-25 17:45:07,142 [Receiver]: <- STX
2023-09-25 17:45:07,144 Seq1: 22
2023-09-25 17:45:07,145 Seq2: 22
2023-09-25 17:45:07,146 Munched data on 1026 byte read, b"\xc7q...." to b"0q..."
2023-09-25 17:45:07,149 [Receiver]: CRC verification failed. Sender: a7cb, Receiver: f8cf.
2023-09-25 17:45:07,151 [Receiver]: Checksum failed.
2023-09-25 17:45:07,152 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:09,157 [Receiver]: NAK -> c=None
2023-09-25 17:45:09,171 Seq1: 2
2023-09-25 17:45:09,172 Seq2: 233
2023-09-25 17:45:09,173 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:09,175 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:11,180 [Receiver]: NAK -> c=None
2023-09-25 17:45:11,189 Seq1: 2
2023-09-25 17:45:11,190 Seq2: 233
2023-09-25 17:45:11,191 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:11,193 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:13,199 [Receiver]: NAK -> c=None
2023-09-25 17:45:13,218 Seq1: 2
2023-09-25 17:45:13,219 Seq2: 233
2023-09-25 17:45:13,220 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:13,222 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:15,227 [Receiver]: NAK -> c=None
2023-09-25 17:45:15,236 Seq1: 2
2023-09-25 17:45:15,237 Seq2: 233
2023-09-25 17:45:15,238 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:15,240 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:17,245 [Receiver]: NAK -> c=None
2023-09-25 17:45:17,257 Seq1: 2
2023-09-25 17:45:17,258 Seq2: 233
2023-09-25 17:45:17,258 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:17,262 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:19,265 [Receiver]: NAK -> c=None
2023-09-25 17:45:19,279 Seq1: 2
2023-09-25 17:45:19,280 Seq2: 233
2023-09-25 17:45:19,280 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:19,284 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:21,287 [Receiver]: NAK -> c=None
2023-09-25 17:45:21,302 Seq1: 2
2023-09-25 17:45:21,303 Seq2: 233
2023-09-25 17:45:21,303 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:21,307 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:23,311 [Receiver]: NAK -> c=None
2023-09-25 17:45:23,327 Seq1: 2
2023-09-25 17:45:23,328 Seq2: 233
2023-09-25 17:45:23,329 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:23,331 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:25,337 [Receiver]: NAK -> c=None
2023-09-25 17:45:25,345 Seq1: 2
2023-09-25 17:45:25,347 Seq2: 233
2023-09-25 17:45:25,348 [Receiver]: Wrong sequence (2, 233, 22), drop the whole packet.
2023-09-25 17:45:25,350 [Receiver]: Send a request for retransmission.
2023-09-25 17:45:27,355 [Receiver]: NAK -> c=None
2023-09-25 17:45:27,373 Seq1: 24
2023-09-25 17:45:27,373 Seq2: 231
2023-09-25 17:45:27,374 [Receiver]: Wrong sequence (24, 231, 22), drop the whole packet.
2023-09-25 17:45:28,376 [Receiver]: The number of retransmissions has reached the maximum limit, abort and exit!
2023-09-25 17:45:28,377 [Receiver]: CAN ->
```

This leads me to wonder if, on line 808 where the full packet is read to then discard, maybe that line is reading too long and eating the sequence byte.

**Comment by alexwhittemore** [comment]
Ahh, I think I fixed it! The smoking gun was `[Receiver]: NAK -> c=None`

I'd changed line 814 here

https://github.com/alexwoo1900/ymodem/blob/8cf009b62f89ef1ae36416d9b4d3260404bf789b/ymodem/Socket.py#L809-L815

to read
``` python
self.logger.debug(f"[Receiver]: NAK -> c={c}")
```

Of course, if that c is coming up `None`, that implies the "wait" isn't waiting long enough to actually get a byte. That call returns too soon with nothing to show for it, and since the transmitter DOES eventually send a byte, it causes a subsequent off-by-one error. 

I'm not sure what the defining characteristic is for that timeout, but I changed line 813 to `c = self._write_and_wait(NAK, [SOH, STX, CAN], 5)` with a 5s timeout, and now my link can recover from CRC errors correctly!

**Comment by alexwoo1900** [comment]
@alexwhittemore Thanks for your test log. 

Yes, this protocol has very high requirement for timing. Users in different environments usually need to adjust timeout events based on their own situation.

I am not sure if there is a need to redo testing or modify the code for this issue, I need more user feedback on such issues.

**Comment by alexwhittemore** [comment]
Yeah, I think I agree. If I were going to demand any sort of change, I think it'd be along the lines of adding some antagonistic test cases in the demos, and providing users some kind of mechanism to trivially adjust/fiddle with timeouts. That one that I found was certainly so buried in the code a reasonable person might assume it's simply broken. It does seem reasonable to me that where such a parameter might have a major impact on real-world function, it'd be flagged in docs and easily adjustable. 

If you or anyone else is interested, I just came up with this final test case to stress the system a bit:

``` python
    # Create up to 5 failures with a 1-in-100 chance on any read.
    failcount = 0
    def read(size, timeout = 10):
        global failcount
        serial_io.timeout = timeout
        if failcount<5 and random.randint(0,100) == 5:
            failcount +=1
            data_in = serial_io.read(size)
            data = bytearray(data_in)
            data[0] = ord('0')
            log.debug(f"Munched data on {size} byte read, failcount {failcount}, {data_in} to {bytes(data)}")
            return bytes(data)
        else:
            return serial_io.read(size)
```

Instead of arbitrarily injecting a single mangled byte, it'll stochastically insert up to 5 errors. Running it a few times on the `sample.stl` file, it's clear there are some errors that aren't recoverable and result in ungraceful disconnect. But on the other hand, I also had many runs where all 5 errors were successfully detected and corrected, and a hash-matching file was finally transmitted. So all in all I think this makes the library viable in my application, and I deeply thank you for all your work!

**Comment by alexwhittemore** [comment]
Ahh, I do seem to be encountering a failure now that might indicate room for improvement. The code includes

```python
'''
7.3.2 Receive_Program_Considerations

Synchronizing: If a valid block number is received, it will be: 
1) the expected one, in which case everything is fine;
2) a repeat of the previously received block. This should be considered OK, 
and only indicates that the receivers <ack> got glitched, and the sender retransmitted; 
3) any other block number indicates a fatal loss of synchronization, such as the rare case
of the sender getting a line-glitch that looked like an <ack>. Abort the transmission, sending a <can>
'''
```

I think I'm actually running into (2) right now, which isn't actually handled gracefully. The receiver sees (remember, my custom debug modifications): 
```
2023-09-25 19:06:05,294 [Receiver]: Wrong sequence (13, 13, 14), drop the whole packet.
2023-09-25 19:06:05,296 [Receiver]: Send a request for retransmission.
2023-09-25 19:06:06,358 [Receiver]: NAK ->
2023-09-25 19:06:06,360 [Receiver]: <- CAN
``` 
That implies block 13 was successfully received and written, but what got smashed on was the ACK back to the transmitter. Over there, we see
```
2023-09-25 19:06:01,151 [Sender]: Data packet 13 ->
2023-09-25 19:06:03,112 [Sender]: No ACK from Receiver, preparing to retransmit.
2023-09-25 19:06:03,170 [Sender]: Data packet 13 ->
2023-09-25 19:06:04,186 [Sender]: No ACK from Receiver, preparing to retransmit.
```
so it looks like there needs to be some logic to say "`seq1` and `seq2` agree and are one-behind `sequence`, so let's just let the transmitter send its block and ACK it while actually throwing it away"

**Comment by alexwoo1900** [comment]
@alexwhittemore What's your device for the testing? Can you post your read and write function here?

**Comment by alexwoo1900** [comment]
Will check in a few days

**Comment by alexwoo1900** [comment]
fixed 7.3.2 and will add some test codes later.

**Comment by alexwhittemore** [comment]
Thanks!
